airflow/models.py

```diff
@@ -1288,19 +1288,19 @@ def run(
             verbose=True)
 
         if not runnable and not mark_success:
-            if self.state != State.QUEUED:
-                # If a task's dependencies are met but it can't be run yet then queue it
-                # instead
-                self.state = State.QUEUED
-                msg = "Queuing attempt {attempt} of {total}".format(
-                    attempt=self.try_number % (task.retries + 1) + 1,
-                    total=task.retries + 1)
-                logging.info(hr + msg + hr)
-
-                self.queued_dttm = datetime.now()
-                msg = "Queuing into pool {}".format(self.pool)
-                logging.info(msg)
-                session.merge(self)
+            # we might have hit concurrency limits, which means we probably
+            # have been running prematurely.
+            self.state = State.UP_FOR_RETRY
```

Comment:
@saguziel Agree on the concurrency note, although I think that is logged (although maybe not clear enough):

```
        runnable = self.are_dependencies_met(
            dep_context=dep_context,
            session=session,
            verbose=True)
```

I understand why you consider QUEUD not being and end state here, but I would argue we would need to set the TaskInstance to RUNNING asap (we have picked it up from the Executor) and then find out if we should be running at all (eg. change to skipped, whatever else).

Also SCHEDULED does not consider the back off period, which I think is required and is not addressed by your (plural :) ) comments.